### DISCUSSION TITLE
Why does adding property to Vertex is slower than Edge?

### DISCUSSION THREAD

**Discussion by CysOrion** [JanusGraph/janusgraph#4342] [Q&A]
I was performing `addPropertySteps` following `addV` and `addE` steps, no index is linked to the records. I used the default traversal source `g`. All properties are `STRING` properties with `SINGLE` cardinality.

Here are the test results:

```

g.addV("person").property("Name","John").profile()
==>Traversal Metrics
Step                                                               Count  Traversers       Time (ms)    % Dur
=============================================================================================================
AddVertexStartStep({label=[person], Name=[John]})                      1           1        2444.200   100.00
                                            >TOTAL                     -           -        2444.200        -


g.addV("person").property("name", "Kyle").profile()
==>Traversal Metrics
Step                                                               Count  Traversers       Time (ms)    % Dur
=============================================================================================================
AddVertexStartStep({name=[Kyle], label=[person]})                      1           1         260.520   100.00
                                            >TOTAL                     -           -         260.520        -



g.addE("Relationship").from(__.V(4240)).to(__.V(8336)).property("related_to", "Son of").profile()
==>Traversal Metrics
Step                                                               Count  Traversers       Time (ms)    % Dur
=============================================================================================================
AddEdgeStartStep({related_to=[Son of], ~from=[[...                     1           1           4.649   100.00
  GraphStep(vertex,[4240])                                             1           1           0.045
  GraphStep(vertex,[8336])                                             1           1           0.089
                                            >TOTAL                     -           -           4.649        -


```

The first `addV` is slower because it is the first traversal performed, but the followed `addV` is still much slower then `addE`'s. Is this a expected behavior?

**Comment by li-boxuan** [comment]
In theory, adding a vertex shall be faster than adding an edge because an edge needs two writes while a vertex needs only one.

I noticed in your second addV query, you used property key "name" while in your first query you used "Name". They are two different properties (IIRC JanusGraph is case-sensitive). I suspect your second query was slow because it needed to update the schema.

**Reply by CysOrion** [reply]
@li-boxuan  Thanks for pointing out the typo, that explains the slower `addV`. Maybe I didn't frame the question properly. My question is why adding property to edge is much faster?

I tested only addV and addE both are fast. But when appended .property() step on both, addV is slower, just like the test I posted! Is there a reason behind this?

**Reply by li-boxuan** [reply]
> I tested only addV and addE both are fast. But when appended .property() step on both, addV is slower

I am surprised to hear that. Are you sure you've done enough experiments that consistently show this behaviour? Ideally, you should use JMH to benchmark.
